toolchain/check/context.cpp

```diff
@@ -118,6 +118,38 @@ auto Context::FinishInst(SemIR::InstId inst_id, SemIR::Inst inst) -> void {
   }
 }
 
+// Returns whether a parse node associated with an instruction of kind
+// `src_kind` is known to be usable as the location of an instruction of kind
+// `target_kind`. This may have false negatives for cases that don't occur in
+// practice.
+static auto HasKnownCompatibleNodeKind(SemIR::InstKind src_kind,
+                                       SemIR::InstKind target_kind) -> bool {
+  if (src_kind == target_kind) {
+    return true;
+  }
+  if (src_kind == SemIR::ImportDecl::Kind &&
+      target_kind == SemIR::Namespace::Kind) {
+    static_assert(
+        std::is_convertible_v<decltype(SemIR::ImportDecl::Kind)::TypedNodeId,
+                              decltype(SemIR::Namespace::Kind)::TypedNodeId>);
+    return true;
+  }
+  return false;
+}
+
+auto Context::CheckCompatibleImportedNodeKind(
+    SemIR::ImportIRInstId imported_loc_id, SemIR::InstKind kind) -> void {
+  auto& import_ir_inst = import_ir_insts().Get(imported_loc_id);
+  auto imported_kind = import_irs()
+                           .Get(import_ir_inst.ir_id)
+                           .sem_ir->insts()
+                           .Get(import_ir_inst.inst_id)
+                           .kind();
```

Comment:
Is this something that could be broken apart to improve wrapping? e.g.:

```suggestion
  const auto* import_ir = import_irs().Get(import_ir_inst.ir_id).sem_ir;
  auto imported_kind = import_ir->insts().Get(import_ir_inst.inst_id).kind();
```